File: 21-MergeTwoSortedLists.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} list1
 * @param {ListNode} list2
 * @return {ListNode}
 */
var mergeTwoLists = function (list1, list2) {

    if(!list1&&!list2){
        return null;
    }
  let ans = new ListNode();  //ans head node
  let ptr = ans;  //ptr to keep track of where we are in our answers list


  while (list1|| list2) {
    let nextNode = new ListNode();
    //next 4 if cases

    //both pointers exist and list1 smaller
    if (list1 && list2 && list1.val <= list2.val) {
      ptr.val = list1.val;
      list1 = list1.next;
    } 
    //both pointers exist and list2 smaller
    else if (list1 && list2 && list2.val < list1.val) {
      ptr.val = list2.val;
      list2 = list2.next;
    } 
    //list 1 exists and list 2 is done
    else if (list1 && !list2) {
      ptr.val = list1.val;
      list1 = list1.next;
    } 
    //list 2 exists and list1 is done
    else if (list2 && !list1) {
      ptr.val = list2.val;
      list2 = list2.next;
    }

    //check if need a new node at the end or if we are done
    if(list1 ||list2 ){
    ptr.next = nextNode;
    ptr = nextNode;
    }
  }
  return ans;
};

File: 21-MergeTwoSortedLists.py
'''
My solution: Parallel iterative approach using 2 pointers and no additional memory
O(N+M) time
O(1) memory


'''


# Definition for singly-linked list.
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # Get the basic edge cases out of the way
        if (list1 is None and list2 is not None):
            return list2
        elif (list1 is not None and list2 is None):
            return list1
        elif (list1 == None and list2 == None):
            return list1

        head = ListNode()
        curr = head
        while (list1 is not None or list2 is not None):
            if (list1 and list2 and list1.val <= list2.val):
                curr.next = list1
                curr = curr.next
                list1 = list1.next
            elif (list1 and list2 and list1.val > list2.val):
                curr.next = list2
                curr = curr.next
                list2 = list2.next

            elif (list1 and not list2):
                curr.next = list1
                curr = curr.next
                list1 = list1.next
            elif (list2 and not list1):
                curr.next = list2
                curr = curr.next
                list2 = list2.next
        return head.next
